r/src/compute.cpp

```diff
@@ -223,4 +205,66 @@ std::shared_ptr<arrow::Table> Table__FilterChunked(
   }
   return tab;
 }
+
+arrow::Datum to_datum(SEXP x) {
+  // TODO: this is repetitive, can we DRY it out?
+  if (Rf_inherits(x, "ArrowObject") && Rf_inherits(x, "Array")) {
+    Rcpp::ConstReferenceSmartPtrInputParameter<std::shared_ptr<arrow::Array>> obj(x);
+    return static_cast<std::shared_ptr<arrow::Array>>(obj);
+  } else if (Rf_inherits(x, "ArrowObject") && Rf_inherits(x, "ChunkedArray")) {
+    Rcpp::ConstReferenceSmartPtrInputParameter<std::shared_ptr<arrow::ChunkedArray>> obj(x);
+    return static_cast<std::shared_ptr<arrow::ChunkedArray>>(obj);
+  } else if (Rf_inherits(x, "ArrowObject") && Rf_inherits(x, "RecordBatch")) {
+    Rcpp::ConstReferenceSmartPtrInputParameter<std::shared_ptr<arrow::RecordBatch>> obj(x);
+    return static_cast<std::shared_ptr<arrow::RecordBatch>>(obj);
+  } else if (Rf_inherits(x, "ArrowObject") && Rf_inherits(x, "Table")) {
+    Rcpp::ConstReferenceSmartPtrInputParameter<std::shared_ptr<arrow::Table>> obj(x);
+    return static_cast<std::shared_ptr<arrow::Table>>(obj);
+  } else {
+    // TODO: scalar?
```

Comment:
Currently elsewhere we use a scalar wrapper function from the dataset namespace (see https://github.com/apache/arrow/blob/master/r/src/expression.cpp#L106-L158). What should we be doing here?

Reply:
I would recommend refactoring that function into `std::shared_ptr<Scalar> to_scalar(SEXP)`. That function need not be exported (same logic as `Datum`; `SEXP` already holds whatever value). The climax of `to_scalar` should be a call to `arrow::MakeScalar`.

After this:
```c++
std::shared_ptr<ds::Expression> dataset___expr__scalar(SEXP x) {
  return ds::scalar(to_scalar(x));
}
```

Reply:
Can we create an R6 class wrapper for `arrow::Scalar`?

Reply:
Sure, we can, and I think we should, unlike Datum, because it makes sense to hold a Scalar in R. Though I suspect that converting from Scalar to R requires some non-trivial refactoring of array-to-vector.cpp.